packages/core/src/components/forms/useAsyncControllableValue.ts

```diff
@@ -0,0 +1,138 @@
+) {
+    const { onCompositionStart, onCompositionEnd, value: propValue, onChange } = props;
+
+    // The source of truth for the input value. This is not updated during IME composition.
+    // It may be updated by a parent component.
+    const [value, setValue] = React.useState(propValue);
+
+    // The latest input value, which updates during IME composition.
+    const [nextValue, setNextValue] = React.useState(propValue);
+
+    // Whether we are in the middle of a composition event.
+    const [isComposing, setIsComposing] = React.useState(false);
+
+    // Whether there is a pending update we are expecting from a parent component.
+    const [hasPendingUpdate, setHasPendingUpdate] = React.useState(false);
+
+    const cancelPendingCompositionEnd = React.useRef<(() => void) | null>(null);
+
+    const handleCompositionStart: React.CompositionEventHandler<E> = React.useCallback(
+        event => {
+            cancelPendingCompositionEnd.current?.();
+            setIsComposing(true);
+            onCompositionStart?.(event);
+        },
+        [onCompositionStart],
+    );
+
+    // creates a timeout which will set `isComposing` to false after a delay
+    // returns a function which will cancel the timeout if called before it fires
+    const createOnCancelPendingCompositionEnd = React.useCallback(() => {
+        const timeoutId = window.setTimeout(
+            () => setIsComposing(false),
+            ASYNC_CONTROLLABLE_VALUE_COMPOSITION_END_DELAY,
+        );
+        return () => window.clearTimeout(timeoutId);
+    }, []);
+
+    const handleCompositionEnd: React.CompositionEventHandler<E> = React.useCallback(
+        event => {
+            // In some non-latin languages, a keystroke can end a composition event and immediately afterwards start another.
+            // This can lead to unexpected characters showing up in the text input. In order to circumvent this problem, we
+            // use a timeout which creates a delay which merges the two composition events, creating a more natural and predictable UX.
+            // `this.state.nextValue` will become "locked" (it cannot be overwritten by the `value` prop) until a delay (10ms) has
+            // passed without a new composition event starting.
+            cancelPendingCompositionEnd.current = createOnCancelPendingCompositionEnd();
+            onCompositionEnd?.(event);
+        },
+        [createOnCancelPendingCompositionEnd, onCompositionEnd],
+    );
+
+    const handleChange: React.ChangeEventHandler<E> = React.useCallback(
+        event => {
+            const { value: targetValue } = event.target;
+            setNextValue(targetValue);
+            onChange?.(event);
+        },
+        [onChange],
+    );
+
+    if (isComposing || propValue === undefined) {
+        // don't derive anything from props if:
+        // - in uncontrolled mode, OR
+        // - currently composing, since we'll do that after composition ends
+    } else {
```

Comment:
How do you feel about this suggestion?

```suggestion
    // don't derive anything from props if:
    // - in uncontrolled mode, OR
    // - currently composing, since we'll do that after composition ends
    const isStateDerivationPaused =  isComposing || propValue === undefined
    if (!isStateDerivationPaused) {
```

It feels like the empty if block could be replaced somehow, but not too opinionated on that. I do see it came from the early return in the `getDerivedStateFromProps` version of this, so I think this is okay to keep around for the hook version.